sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -60,9 +60,9 @@ object JdbcUtils extends Logging with SQLConfHelper {
    * @param options - JDBC options that contains url, table and other information.
    * @throws IllegalArgumentException if the driver could not open a JDBC connection.
    */
-  def createConnectionFactory(options: JDBCOptions): () => Connection = {
+  def createConnectionFactory(options: JDBCOptions): Option[Partition] => Connection = {
```

Comment:
how do you overwrite this in your clickhouse data source?

Reply:
Seems like Spark does not allow JDBC dialects to customize the connection creation yet.

Reply:
We get the jdbc connection looks like:
```
private def createConnectionFactoryByShard(options: JDBCOptions): Partition => Connection = {
    val shards = ShardOptions.create(options)
    val driverClass: String = options.driverClass
    (thePart: Partition) => {
      require(thePart.index < shards.shards.length)
      val url = shards.shards.apply(thePart.index)
      logInfo(s"Create connection for shard: $url")
      DriverRegistry.register(driverClass)
      val driver: Driver = DriverManager.getDrivers.asScala.collectFirst {
        case d: DriverWrapper if d.wrapped.getClass.getCanonicalName == driverClass => d
        case d if d.getClass.getCanonicalName == driverClass => d
      }.getOrElse {
        throw new IllegalStateException(
          s"Did not find registered driver with class $driverClass")
      }
      driver.connect(url, options.asConnectionProperties)
    }
  }
```